core/src/impl/Kokkos_Half_MathematicalFunctions.hpp

```diff
@@ -228,6 +233,87 @@ KOKKOS_INLINE_FUNCTION bool isnan(Kokkos::Experimental::bhalf_t x) {
       ((bit_pattern_x.value & fraction_mask.value) != 0));
 }
 #endif
+
+#if !(defined(KOKKOS_ENABLE_CUDA) && defined(KOKKOS_COMPILER_MSVC))
+template <typename fp16_t>
+KOKKOS_INLINE_FUNCTION fp16_t nextafter_half_impl(fp16_t from, fp16_t to) {
+  static_assert((std::is_same_v<fp16_t, Kokkos::Experimental::half_t> ||
+                 std::is_same_v<fp16_t, Kokkos::Experimental::bhalf_t>)
+                 && sizeof(fp16_t) == 2, "nextafter_impl only supports half_t and bhalf_t");
+  constexpr std::uint16_t FP16_SIGN_MASK = 0x8000;
+  constexpr std::uint16_t FP16_POS_ZERO  = 0x0000;
+  constexpr std::uint16_t FP16_NEG_ZERO  = 0x8000;
+  constexpr std::uint16_t FP16_SMALLEST_POS_DN = 0x0001;  // Smallest positive denormal
+  constexpr std::uint16_t FP16_SMALLEST_NEG_DN = 0x8001;  // Smallest negative denormal (magnitude)
+
+   // Handle Nans
+   if (isnan(from) || isnan(to)) {
+     return Kokkos::Experimental::quiet_NaN<fp16_t>::value;
+   }
+
+   // Handle equality
+   if (from == to) return to;
+
+   // Get unsigned integer representation of from
+   std::uint16_t uint_from = bit_cast<std::uint16_t, fp16_t>(from);
+
+   // Handle zeros
+   if (uint_from == FP16_POS_ZERO || uint_from == FP16_NEG_ZERO) {
+     // from is +0.0 or -0.0
+     // Return smallest magnitude number with the sign of 'to'.
+     // nextafter(±0, negative) -> smallest_negative
+     // nextafter(±0, positive) -> smallest_positive
+     return bit_cast<fp16_t, std::uint16_t>((to > from) ? FP16_SMALLEST_POS_DN
+                                                        : FP16_SMALLEST_NEG_DN);
+   }
+
+   // Determine direction and sign of 'from'
+   // True if moving to positive infinity
+   bool to_positive_infinity = (to > from);
+   bool from_is_negative     = ((uint_from & FP16_SIGN_MASK) != 0);
+
+   std::uint16_t uint_result;
+   if (from_is_negative) {
+     // For negative numbers, increasing magnitude means moving towards -inf
+     // (larger uint value) Decreasing magnitude means moving towards zero
+     // (smaller uint value)
+     if (to_positive_infinity) {
+       // Moving toward zero or positive
+       uint_result = uint_from - 1;
+     } else {
+       // Moving toward negative infinity
+       uint_result = uint_from + 1;
+     }
+   } else {
+     // For positive numbers, increasing magnitude means moving towards +inf
+     // (larger uint value) Decreasing magnitude means moving towards zero
+     // (smaller uint value)
+     if (to_positive_infinity) {
+       // Moving toward positive infinity
+       uint_result = uint_from + 1;
+     } else {
+       // Moving toward zero or negative infinity
+       uint_result = uint_from - 1;
+     }
```

Comment:
Do I dare suggesting we could just return `uint_from + to_positive_infinity ^ from_is_negative`?

Reply:
This may work? 
`uint_result = uint_from + 2 * (to_positive_infinity ^ from_is_negative) - 1;`
At least, I would like to keep the original logic in comments for better understanding